File: Sources/SwctxCore/Trace.swift
import Foundation
import GRDB

/// `trace_lookup` — stack trace in, indexed frames out. Parses the
/// common traceback formats (Python, JS/TS `at`, Go panic, generic
/// `path:line`), suffix-matches paths against indexed files, and maps
/// each frame to its enclosing chunk. Pure read; static only — inlined
/// calls, source maps and native frames are not recoverable.
public enum Trace {

    public struct Frame {
        public var raw: String
        public var path: String      // path as written in the trace
        public var line: Int
        public var symbolHint: String? // function name if the format carries one
    }

    // Python:  File "src/x.py", line 42, in fn
    private static let pyRx = try! NSRegularExpression(
        pattern: #"File "([^"]+)", line (\d+)(?:, in (\S+))?"#)
    // Swift/ObjC runtime: Fatal error: …: file /abs/x.swift, line 448
    private static let swiftRx = try! NSRegularExpression(
        pattern: #"\bfile\s+([^",\s=]+\.[A-Za-z0-9]+),\s*line\s+(\d+)"#,
        options: [.caseInsensitive])
    // JS/TS: at fn (src/x.ts:10:5) | at async fn (…) | at src/x.ts:10:5
    private static let jsRx = try! NSRegularExpression(
        pattern: #"\bat\s+(?:async\s+)?(?:([\w$.<>]+)\s+(?:\[as\s+\S+\]\s+)?\()?((?:file://)?[^\s()]+?):(\d+)(?::\d+)?\)?"#)
    // Go panic:   /path/file.go:37 +0x1f2   (function on previous line)
    private static let goRx = try! NSRegularExpression(
        pattern: #"^\s*(\S+\.go):(\d+)\s+\+0x[0-9a-fA-F]+"#, options: [.anchorsMatchLines])
    // Generic fallback: any `something.ext:N` — Swift/Java/Rust/C++/etc.
    private static let genericRx = try! NSRegularExpression(
        pattern: #"([A-Za-z0-9_./\\-]+\.(?:swift|py|js|jsx|ts|tsx|mjs|cjs|go|rs|java|kt|rb|php|m|mm|c|cc|cpp|h|hpp)):(\d+)"#)

    /// Parse frames in textual order. Specific formats run first; the
    /// generic `path:N` fallback skips ranges already claimed by them.
    public static func parse(_ text: String) -> [Frame] {
        var frames: [(Int, Frame)] = []
        var claimed: [NSRange] = []
        func add(_ rx: NSRegularExpression, pathAt: Int, lineAt: Int,
                 symAt: Int) {
            let ns = text as NSString
            for m in rx.matches(in: text, range: NSRange(text.startIndex..., in: text)) {
                if claimed.contains(where: {
                    NSIntersectionRange($0, m.range).length > 0 }) { continue }
                claimed.append(m.range)
                let path = m.range(at: pathAt).location != NSNotFound
                    ? ns.substring(with: m.range(at: pathAt)) : ""
                let line = m.range(at: lineAt).location != NSNotFound
                    ? Int(ns.substring(with: m.range(at: lineAt))) ?? 0 : 0
                let sym = symAt > 0 && m.range(at: symAt).location != NSNotFound
                    ? ns.substring(with: m.range(at: symAt)) : nil
                guard !path.isEmpty, line > 0 else { continue }
                frames.append((m.range.location,
                    Frame(raw: ns.substring(with: m.range),
                          path: path, line: line, symbolHint: sym)))
            }
        }
        add(pyRx, pathAt: 1, lineAt: 2, symAt: 3)
        add(swiftRx, pathAt: 1, lineAt: 2, symAt: -1)
        add(jsRx, pathAt: 2, lineAt: 3, symAt: 1)
        add(goRx, pathAt: 1, lineAt: 2, symAt: -1)
        add(genericRx, pathAt: 1, lineAt: 2, symAt: -1)
        return frames.sorted { $0.0 < $1.0 }.map { $0.1 }
    }

    /// Trace paths are often absolute/container paths (/app/…, site-
    /// packages, node_modules). Match workspace files by longest path
    /// suffix; frames that match nothing are returned unmatched.
    public static func resolve(store: Store, frames: [Frame])
        -> [[String: Any]] {
        let paths: [String] = (try? store.pool.read { db in
            try String.fetchAll(db, sql: "SELECT path FROM files ORDER BY path")
        }) ?? []
        var out: [[String: Any]] = []
        for f in frames {
            let norm = f.path
                .replacingOccurrences(of: "file://", with: "")
                .replacingOccurrences(of: "\\", with: "/")
            // Best suffix match: longest indexed path that ends the
            // trace path, or that the trace path ends with.
            var best: String?
            for p in paths {
                if norm == p || norm.hasSuffix("/" + p) || p.hasSuffix("/" + norm) {
                    if (best?.count ?? 0) < p.count { best = p }
                } else {
                    // component-wise suffix: "…/src/x.py" vs "src/x.py"
                    let t = norm.split(separator: "/"), ip = p.split(separator: "/")
                    if ip.count <= t.count,
                       Array(t.suffix(ip.count)) == ip,
                       (best?.count ?? 0) < p.count { best = p }
                }
            }
            var rec: [String: Any] = [
                "raw": f.raw, "trace_path": f.path, "line": f.line,
                "matched": best != nil]
            if let sym = f.symbolHint { rec["symbol_hint"] = sym }
            if let p = best {
                rec["path"] = p
                let row = try? store.pool.read({ db in
                    try Row.fetchOne(db, sql: """
                        SELECT c.id, c.symbol, c.start_line, c.end_line
                        FROM chunks c JOIN files f ON f.id = c.file_id
                        WHERE f.path = ? AND c.start_line <= ? AND c.end_line >= ?
                        ORDER BY (c.end_line - c.start_line) ASC LIMIT 1
                        """, arguments: [p, f.line, f.line])
                })
                if let row, let cid = row["id"] as? Int64 {
                    rec["chunk_id"] = cid
                    rec["span"] = "\((row["start_line"] as? Int64) ?? 0)-\((row["end_line"] as? Int64) ?? 0)"
                }
                // A decl nested inside the matched chunk (method inside a
                // class-sized chunk) is more precise than the chunk's own
                // symbol; for symbol-less `window` chunks the nearest decl
                // above is the owning declaration. Also covers matched
                // lines that land between chunks entirely.
                let decl = try? store.pool.read({ db in
                    try Simulate.nearestDecl(db, path: p, line: Int64(f.line))
                })
                if let enc = Simulate.enclosingSymbol(
                    chunkSymbol: row?["symbol"] as? String,
                    chunkStart: (row?["start_line"] as? Int64) ?? Int64(f.line),
                    decl: decl) {
                    rec["symbol"] = enc.0
                    rec["symbol_source"] = enc.1
                }
            }
            out.append(rec)
        }
        return out
    }

    /// Deepest matched app frame is the crash site; suspects are the indexed
    /// callers of that frame's symbol. Looked up by `dst_name` (not only
    /// `dst_chunk`): crash lines often land in symbol-less `window` chunks
    /// that no edge points at, and name-only edges carry unresolved callers
    /// too. `resolved` marks edges pinned to a same-named definition.
    static func suspects(store: Store, resolved: [[String: Any]])
        -> [[String: Any]] {
        guard let last = resolved.last(where: { ($0["matched"] as? Bool) == true })
        else { return [] }
        let sym = last["symbol"] as? String
        let cid = last["chunk_id"] as? Int64
        guard sym != nil || cid != nil else { return [] }
        var defs = Set<Int64>()
        let rows: [Row] = (try? store.pool.read { db in
            if let sym {
                defs = Set(try Int64.fetchAll(db, sql:
                    "SELECT chunk_id FROM symbols WHERE name = ?",
                    arguments: [sym]))
                return try Row.fetchAll(db, sql: """
                    SELECT DISTINCT f.path, e.line, c.symbol, e.dst_chunk
                    FROM edges e
                    JOIN chunks c ON c.id = e.src_chunk
                    JOIN files f ON f.id = c.file_id
                    WHERE e.dst_name = ? AND e.kind IN
                        ('calls','instantiates','uses_type','api_call')
                    ORDER BY f.path, e.line LIMIT 20
                    """, arguments: [sym])
            }
            return try Row.fetchAll(db, sql: """
                SELECT DISTINCT f.path, e.line, c.symbol, e.dst_chunk
                FROM edges e
                JOIN chunks c ON c.id = e.src_chunk
                JOIN files f ON f.id = c.file_id
                WHERE e.dst_chunk = ? AND e.kind IN
                    ('calls','instantiates','uses_type','api_call')
                ORDER BY f.path, e.line LIMIT 20
                """, arguments: [cid ?? -1])
        }) ?? []
        // Files touched by the most recent commits are statistically the
        // likelier break source — flag suspects on that list. Commit
        // records carry a `files` array in payload; LIKE-match is cheap.
        let recentPaths: Set<String> = (try? store.pool.read { db in
            let payloads = try String.fetchAll(db, sql: """
                SELECT payload FROM records WHERE kind = 'commit'
                ORDER BY id DESC LIMIT 20
                """)
            var set = Set<String>()
            for pl in payloads {
                guard let data = pl.data(using: .utf8),
                      let obj = try? JSONSerialization.jsonObject(with: data)
                        as? [String: Any],
                      let ps = obj["files"] as? [String] else { continue }
                set.formUnion(ps)
            }
            return set
        }) ?? []
        return rows.map {
            let p = ($0["path"] as? String) ?? ""
            var r: [String: Any] = [
                "path": p,
                "line": ($0["line"] as? Int64) ?? 0,
                "symbol": ($0["symbol"] as? String) ?? NSNull(),
                "resolved": ($0["dst_chunk"] as? Int64)
                    .map { defs.contains($0) } ?? false]
            if recentPaths.contains(p) { r["recent_commit"] = true }
            return r
        }
        // Recently commit-touched callers triage first.
        .sorted { a, b in
            ((a["recent_commit"] as? Bool) == true) != ((b["recent_commit"] as? Bool) == true)
                ? (a["recent_commit"] as? Bool) == true
                : (a["path"] as? String ?? "") < (b["path"] as? String ?? "")
        }
    }
}

File: swctx-py/swctx_py/trace.py
"""trace_lookup — stack trace in, indexed frames out (port of
SwctxCore/Trace.swift). Pure read over files/chunks/symbols/edges.

Parses Python `File "p", line N, in fn`, JS/TS `at fn (p:N:C)`
(incl. `at async`, `[as alias]`), Go panic `p:N +0x…`, and a generic
`path.ext:N` fallback; suffix-matches trace paths to indexed files and
maps each frame to its enclosing chunk. Static only — inlined calls,
source maps and native frames are not recoverable.
"""
from __future__ import annotations

import json as _json
import re

from .store import Store

_PY_RX = re.compile(r'File "([^"]+)", line (\d+)(?:, in (\S+))?')
_JS_RX = re.compile(
    r"\bat\s+(?:async\s+)?"
    r"(?:([\w$.<>]+)\s+(?:\[as\s+\S+\]\s+)?\()?"
    r"((?:file://)?[^\s()]+?):(\d+)(?::\d+)?\)?")
_GO_RX = re.compile(r"^\s*(\S+\.go):(\d+)\s+\+0x[0-9a-fA-F]+", re.M)
_GENERIC_RX = re.compile(
    r"([A-Za-z0-9_./\\-]+\.(?:swift|py|js|jsx|ts|tsx|mjs|cjs|go|rs|java"
    r"|kt|rb|php|m|mm|c|cc|cpp|h|hpp)):(\d+)")


def parse(text: str) -> list[dict]:
    """Frames in textual order; generic fallback skips ranges already
    claimed by the specific formats."""
    frames: list[tuple[int, dict]] = []
    claimed: list[tuple[int, int]] = []

    def add(rx: re.Pattern, path_at: int, line_at: int, sym_at: int):
        for m in rx.finditer(text):
            if any(a <= m.start() < b or m.start() <= a < m.end()
                   for a, b in claimed):
                continue
            claimed.append(m.span())
            path = m.group(path_at) or ""
            try:
                line = int(m.group(line_at))
            except (TypeError, ValueError):
                line = 0
            sym = m.group(sym_at) if sym_at > 0 else None
            if not path or line <= 0:
                continue
            frames.append((m.start(), {
                "raw": m.group(0), "path": path, "line": line,
                "symbol_hint": sym}))

    add(_PY_RX, 1, 2, 3)
    add(_JS_RX, 2, 3, 1)
    add(_GO_RX, 1, 2, -1)
    add(_GENERIC_RX, 1, 2, -1)
    return [f for _, f in sorted(frames, key=lambda t: t[0])]


def resolve(store: Store, frames: list[dict]) -> list[dict]:
    paths = sorted(r[0] for r in store.db.execute("SELECT path FROM files"))
    out: list[dict] = []
    for f in frames:
        norm = f["path"].replace("file://", "").replace("\\", "/")
        tparts = norm.split("/")
        best = None
        for p in paths:
            if norm == p or norm.endswith("/" + p) or p.endswith("/" + norm):
                if best is None or len(p) > len(best):
                    best = p
            else:
                iparts = p.split("/")
                if (len(iparts) <= len(tparts)
                        and tparts[-len(iparts):] == iparts
                        and (best is None or len(p) > len(best))):
                    best = p
        rec = {"raw": f["raw"], "trace_path": f["path"],
               "line": f["line"], "matched": best is not None}
        if f.get("symbol_hint"):
            rec["symbol_hint"] = f["symbol_hint"]
        if best:
            rec["path"] = best
            row = store.db.execute(
                "SELECT id, symbol_name, start_line, end_line FROM chunks "
                "WHERE file_id = ? AND start_line <= ? AND end_line >= ? "
                "ORDER BY (end_line - start_line) ASC LIMIT 1",
                (best, f["line"], f["line"])).fetchone()
            if row:
                rec["chunk_id"] = row[0]
                rec["symbol"] = row[1]
                rec["span"] = f"{row[2]}-{row[3]}"
        out.append(rec)
    return out


def suspects(store: Store, resolved: list[dict]) -> list[dict]:
    last = next((r for r in reversed(resolved) if r.get("matched")), None)
    if not last or "chunk_id" not in last:
        return []
    rows = store.db.execute(
        "SELECT DISTINCT c.file_id, e.line, c.symbol_name "
        "FROM edges e JOIN chunks c ON c.id = e.src_chunk "
        "WHERE e.dst_chunk = ? AND e.kind IN "
        "('calls','instantiates','uses_type','api_call') "
        "ORDER BY c.file_id LIMIT 20", (last["chunk_id"],)).fetchall()
    recent: set[str] = set()
    for (body,) in store.db.execute(
            "SELECT payload FROM records WHERE kind='commit' "
            "ORDER BY id DESC LIMIT 20"):
        try:
            recent.update(_json.loads(body).get("files", []))
        except (ValueError, AttributeError):
            pass
    out = []
    for path, line, sym in rows:
        r = {"path": path, "line": line, "symbol": sym}
        if path in recent:
            r["recent_commit"] = True
        out.append(r)
    return out


def run(store: Store, trace: str) -> dict:
    frames = parse(trace)
    if not frames:
        return {"frames": [], "count": 0,
                "note": "no frames parsed — expected Python/JS/Go "
                        "traceback lines or path:line references"}
    resolved = resolve(store, frames)
    matched = [r for r in resolved if r["matched"]]
    return {
        "frames": resolved, "count": len(resolved),
        "matched": len(matched),
        "suspects": suspects(store, resolved),
        "note": "static mapping — inlined calls, source maps and "
                "native frames are not recoverable"}
